charts/scalardb/templates/scalardb/deployment.yaml

```diff
@@ -72,6 +73,11 @@ spec:
             value: "{{ .Values.scalardb.storageConfiguration.storage }}"
           - name: SCALAR_DB_LOG_LEVEL
             value: "{{ .Values.scalardb.storageConfiguration.dbLogLevel }}"
+          {{- if .Values.scalardb.secretName }}
+          envFrom:
```

Comment:
What if the same environment variables are defined in `env` and `envFrom`? It seems that their priority is not clear. I googled a little but could not find the specs. Does `envFrom` have precedence over `env` always, or does the order matter?

Reply:
Thank you for your review!

Indeed, I am concerned about priorities.

I tried to test using the same (duplicated) environment variable name.
As a result, the values in `env` are set as environment variables in the Pod, even if I changed the order of `env` and `envFrom`.
So, it seems that the priority does not order matter.

However, I don't find the information in the perspective of specification of Kubernetes yet.
I continue to investigate about priority between `env` and `envFrom`.

---

### Current chart

1. Create `Secret` includes duplicated environment variables.
   ```console
   kubectl create secret generic scalardb-customized-secret \
     --from-literal=SCALAR_DB_CONTACT_POINTS=duplicate_test_contact_point \
     --from-literal=SCALAR_DB_CONTACT_PORT=duplicate_test_port \
     --from-literal=SCALAR_DB_USERNAME=duplicate_test_username \
     --from-literal=SCALAR_DB_PASSWORD=duplicate_test_password \
     --from-literal=SCALAR_DB_STORAGE=duplicate_test_storage \
     --from-literal=SCALAR_DB_LOG_LEVEL=duplicate_test_log_level
   ```
   ```yaml
   $ kubectl get secret scalardb-customized-secret -o yaml
   apiVersion: v1
   data:
     SCALAR_DB_CONTACT_POINTS: ZHVwbGljYXRlX3Rlc3RfY29udGFjdF9wb2ludA==
     SCALAR_DB_CONTACT_PORT: ZHVwbGljYXRlX3Rlc3RfcG9ydA==
     SCALAR_DB_LOG_LEVEL: ZHVwbGljYXRlX3Rlc3RfbG9nX2xldmVs
     SCALAR_DB_PASSWORD: ZHVwbGljYXRlX3Rlc3RfcGFzc3dvcmQ=
     SCALAR_DB_STORAGE: ZHVwbGljYXRlX3Rlc3Rfc3RvcmFnZQ==
     SCALAR_DB_USERNAME: ZHVwbGljYXRlX3Rlc3RfdXNlcm5hbWU=
   kind: Secret
   metadata:
     creationTimestamp: "2022-06-07T04:39:23Z"
     name: scalardb-customized-secret
     namespace: default
     resourceVersion: "186180"
     uid: c831b433-d9d0-403f-9c86-109debe6a935
   type: Opaque
   ```
1. Prepare custom values file.
   ```yaml
   scalardb:
   
     image:
       repository: "local/test/scalardb-server"
       tag: "4.0.0-SNAPSHOT"
   
     secretName: "scalardb-customized-secret"
   ```
1. Deploy Scalar DB Server.
   ```console
   helm install hoge ~/github.com/scalar-labs/helm-charts/charts/scalardb/ -f ./scalardb-server.yaml
   ```
1. Confirm `database.properties` file in the Pod.
   ```console
   $ kubectl exec -t hoge-scalardb-78b449c5b6-bt7qr -- cat database.properties | grep -v -e "^\s*#" -e "^\s*$"
   scalar.db.contact_points=cassandra
   scalar.db.contact_port=9042
   scalar.db.username=cassandra
   scalar.db.password=cassandra
   scalar.db.storage=cassandra
   ```
1. Confirm environment variables in the Pod.
   ```console
   $ kubectl exec -t hoge-scalardb-78b449c5b6-bt7qr -- env | grep -e "^SCALAR_DB"
   SCALAR_DB_STORAGE=cassandra
   SCALAR_DB_USERNAME=cassandra
   SCALAR_DB_CONTACT_POINTS=cassandra
   SCALAR_DB_CONTACT_PORT=9042
   SCALAR_DB_LOG_LEVEL=INFO
   SCALAR_DB_PASSWORD=cassandra
   SCALAR_DB_TRANSACTION_MANAGER=consensus-commit
   SCALAR_DB_ISOLATION_LEVEL=
   SCALAR_DB_CONSENSUSCOMMIT_SERIALIZABLE_STRATEGY=
   ```

---

### Move `envFrom` onto `env`

1. Update chart
   ```diff
   $ git diff
   diff --git a/charts/scalardb/templates/scalardb/deployment.yaml b/charts/scalardb/templates/scalardb/deployment.yaml
   index 209b7ec..9928f62 100644
   --- a/charts/scalardb/templates/scalardb/deployment.yaml
   +++ b/charts/scalardb/templates/scalardb/deployment.yaml
   @@ -46,6 +46,11 @@ spec:
              ports:
                - containerPort: 60051
                - containerPort: 8080
   +          {{- if .Values.scalardb.secretName }}
   +          envFrom:
   +          - secretRef:
   +              name: "{{ .Values.scalardb.secretName }}"
   +          {{- end }}
              env:
              - name: SCALAR_DB_CONTACT_POINTS
                value: "{{ .Values.scalardb.storageConfiguration.contactPoints }}"
   @@ -73,11 +78,6 @@ spec:
                value: "{{ .Values.scalardb.storageConfiguration.storage }}"
              - name: SCALAR_DB_LOG_LEVEL
                value: "{{ .Values.scalardb.storageConfiguration.dbLogLevel }}"
   -          {{- if .Values.scalardb.secretName }}
   -          envFrom:
   -          - secretRef:
   -              name: "{{ .Values.scalardb.secretName }}"
   -          {{- end }}
              resources:
                {{- toYaml .Values.scalardb.resources | nindent 12 }}
              livenessProbe:
   ```
1. Prepare custom values file.
   ```yaml
   scalardb:
   
     image:
       repository: "local/test/scalardb-server"
       tag: "4.0.0-SNAPSHOT"
   
     secretName: "scalardb-customized-secret"
   ```
1. Confirm `Secret`.
   ```yaml
   $ kubectl get secret scalardb-customized-secret -o yaml
   apiVersion: v1
   data:
     SCALAR_DB_CONTACT_POINTS: ZHVwbGljYXRlX3Rlc3RfY29udGFjdF9wb2ludA==
     SCALAR_DB_CONTACT_PORT: ZHVwbGljYXRlX3Rlc3RfcG9ydA==
     SCALAR_DB_LOG_LEVEL: ZHVwbGljYXRlX3Rlc3RfbG9nX2xldmVs
     SCALAR_DB_PASSWORD: ZHVwbGljYXRlX3Rlc3RfcGFzc3dvcmQ=
     SCALAR_DB_STORAGE: ZHVwbGljYXRlX3Rlc3Rfc3RvcmFnZQ==
     SCALAR_DB_USERNAME: ZHVwbGljYXRlX3Rlc3RfdXNlcm5hbWU=
   kind: Secret
   metadata:
     creationTimestamp: "2022-06-07T04:39:23Z"
     name: scalardb-customized-secret
     namespace: default
     resourceVersion: "186180"
     uid: c831b433-d9d0-403f-9c86-109debe6a935
   type: Opaque
   ```
1. Deploy Scalar DB Server.
   ```console
   helm install fuga ~/github.com/scalar-labs/helm-charts/charts/scalardb/ -f ./scalardb-server.yaml
   ```
1. Confirm `database.properties` file in the Pod.
   ```console
   $ kubectl exec -t fuga-scalardb-86c7fc985d-n7xpx -- cat database.properties | grep -v -e "^\s*#" -e "^\s*$"
   scalar.db.contact_points=cassandra
   scalar.db.contact_port=9042
   scalar.db.username=cassandra
   scalar.db.password=cassandra
   scalar.db.storage=cassandra
   ```
1. Confirm environment variables in the Pod.
   ```console
   $ kubectl exec -t fuga-scalardb-86c7fc985d-n7xpx -- env | grep -e "^SCALAR_DB"
   SCALAR_DB_PASSWORD=cassandra
   SCALAR_DB_STORAGE=cassandra
   SCALAR_DB_USERNAME=cassandra
   SCALAR_DB_CONTACT_POINTS=cassandra
   SCALAR_DB_CONTACT_PORT=9042
   SCALAR_DB_LOG_LEVEL=INFO
   SCALAR_DB_TRANSACTION_MANAGER=consensus-commit
   SCALAR_DB_ISOLATION_LEVEL=
   SCALAR_DB_CONSENSUSCOMMIT_SERIALIZABLE_STRATEGY=
   ```

Reply:
I found the docs about `env` and `envFrom`.

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#container-v1-core

* envFrom  
  List of sources to populate environment variables in the container. The keys defined within a source must be a C_IDENTIFIER. All invalid keys will be reported as an event when the container is starting. When a key exists in multiple sources, the value associated with the last source will take precedence. Values defined by an Env with a duplicate key will take precedence. Cannot be updated.

According to this docs, the priority of `env` is high.
If the environment variable names are duplicated between `env` and `envFrom`, the value of `env` will be applied to the Pod.

I will add the note about the above in the document that I'm creating #104.

Reply:
The environment variables like `SCALAR_DB_CONTACT_POINTS` are always defined (even if `.Values.scalardb.storageConfiguration.contactPoints` is null, for example), so isn't it possible for `envFrom` to override these pre-defined variables?
Perhaps should we wrap each `env` variable with `if` statements?

```
          env:
          {{- if Values.scalardb.storageConfiguration.contactPoints }}
          - name: SCALAR_DB_CONTACT_POINTS
            value: "{{ .Values.scalardb.storageConfiguration.contactPoints }}"
          {{- end }}
```

Reply:
> However, the default values are defined in the `values.yaml`.
> So, if we use `{{- if Values.scalardb.storageConfiguration.contactPoints }}` as condition, we need to override its value to null (empty string or 0) explicitly in the custom values file for remove pre-defined environment variables.

I see. I missed that point.

> I think it is better and simple to add restrictions in the document as follows.

That's a good idea.

If we make the variable names reserved, why not change the variable names to be different from the env variables defined in the Dockerfile? Probably they are only used as template variables in the `database.properties` template in the yaml file to substitute the (deprecated) `storageConfiguration` parameters, and they are no longer related to the variables in the Dockerfile.

For example, would it be possible to name these variables as follows? It's easier to understand that the values in the `storageConfiguration` block are substituted in the template since the names are straightforward.

```
    scalar.db.contact_points={{ default .Env.storageConfiguration_contactPoints "" }}
    scalar.db.contact_port={{ default .Env.storageConfiguration_contactPort "" }}
    ...
```

Reply:
> If we make the variable names reserved, why not change the variable names to be different from the env variables defined in the Dockerfile? Probably they are only used as template variables in the `database.properties` template in the yaml file to substitute the (deprecated) `storageConfiguration` parameters, and they are no longer related to the variables in the Dockerfile.

You're right.
These environment variable names are only used in Helm Charts.
And, they are not related to the environment variables in the Dockerfile.

We can separate the environment variable name between Helm Charts and Dockerfile because the Helm Charts completely override the database.properties.

> For example, would it be possible to name these variables as follows? It's easier to understand that the values in the `storageConfiguration` block are substituted in the template since the names are straightforward.

I agree with your suggestion (using Helm Charts original environment variable name).
I will update it. Thank you!

Reply:
@ymorimo  
I updated environment variable names in the `values.yaml` and `deployment.yaml` in 266f2eafb8042f90b21bf7f3036a32b2f75c8382.

Regarding the `scalardb.storageConfiguration.dbLogLevel`, it is NOT passed to the `database.properties`.
It's used in the `log4j.properties.tmpl` in the Dockerfile.
And, the environment variable name (`SCALAR_DB_LOG_LEVEL`) depends on the `log4j.properties.tmpl`.
So, we need to remain the name `SCALAR_DB_LOG_LEVEL` same as environment variable name in the `log4j.properties.tmpl`.

Please take a look again!

---

I confirmed the updated environment variables are passed to the Pod correctly as follows.

* `scalardb-server.yaml`
  ```yaml
  scalardb:
  
    image:
      repository: "local/test/scalardb-server"
      tag: "4.0.0-SNAPSHOT"
  
    storageConfiguration:
      contactPoints: "jdbc:postgresql://postgresql-scalardb.default.svc.cluster.local:5432/postgres"
      username: "postgres"
      password: "postgres"
      storage: "jdbc"
  ```
* `database.properties` in the Pod.
  ```console
  $ kubectl exec -t foo-scalardb-56746689b6-k8rp2 -- cat database.properties | grep -v -e "^\s*#" -e "^\s*$"
  scalar.db.contact_points=jdbc:postgresql://postgresql-scalardb.default.svc.cluster.local:5432/postgres
  scalar.db.contact_port=9042
  scalar.db.username=postgres
  scalar.db.password=postgres
  scalar.db.storage=jdbc
  ```
* Environment variables in the Pod.
  ```console
  $ kubectl exec -t foo-scalardb-56746689b6-k8rp2 -- env | grep storage
  storageConfiguration_contactPort=9042
  storageConfiguration_username=postgres
  storageConfiguration_password=postgres
  storageConfiguration_storage=jdbc
  storageConfiguration_contactPoints=jdbc:postgresql://postgresql-scalardb.default.svc.cluster.local:5432/postgres
  ```

Reply:
Although it is just customary for environment variable names to be in uppercase, I feel that it is strange for environment variable names to be in lowercase. However, there is no problem.

Reply:
> We can separate the environment variable name between Helm Charts and Dockerfile because the Helm Charts completely override the database.properties.

I feel it would be easier to understand if the environment variable names were the same as those used in the configuration file included in the container image. I understand that with this change, the configuration file included in the container image will no longer be used, and the configuration file defined here will be used instead. If that is the case, I think it is better to use the same environment variable name as defined in the container image.

However, there is no problem with changing the names of environment variables.

Reply:
@superbrothers  
Thank you for your comments!
We discussed about environment variable name.
As a result, we decided as follows.

* Use uppercase for environment variable name.
* If the environment variable name (secret key name) of `env` and `envFrom` are the same, the value of `env` will be applied since the priority of `env` is high. So, if we use `SCALAR_DB_XXXX` in the chart, users cannot use `SCALAR_DB_XXXX` as a secret key to pass the credentials since the priority of `env` (internal used `SCALAR_DB_XXXX`) is high. For the reason the above, we want to use another name (original name) in the helm internal.
* These environment variables are used in the Helm internal only, but we (users) can see them in the container after deploying. So, we add the prefix `HELM_` so that users know they are internally used environment variables.
* We need to apply the same updates to another scalar products (Ledger, Auditor, etc...). To make changes in the several charts simple and understandable, we just add prefix `HELM_`.
* We will consider to removing default value of Dockerfile separately in the another PR in the future.